### PR TITLE
Change TensorShape to typically not allocate heap memory

### PR DESCRIPTION
**Pull request by RyanUnderhill**
**Description**: Switch TensorShape from being a std::vector to having a small buffer optimization so that the common cases do not allocate heap memory. Larger shapes will still allocate memory as before

The bulk of the changes are fixing up all code that treats a TensorShape as a std::vector. It's now switched to using a gsl::span as that handles both cases of using the fixed buffer or the dynamically allocated one.

**Motivation and Context**
The allocations from TensorShape are bad for performance as memory allocations are a huge bottleneck. At least one other team has requested we optimize this to reduce the number of memory allocations.

### CODE REVIEW

File: include/onnxruntime/core/framework/tensor_shape.h [hunk 1]
```diff
@@ -16,60 +17,45 @@ namespace onnxruntime {
...
 
-  TensorShape(std::vector<int64_t>&& dims) : std::vector<int64_t>(std::move(dims)) {}
-
-  TensorShape(const std::initializer_list<int64_t>& dims) : std::vector<int64_t>(dims) {}
-
-  TensorShape(const int64_t* dimension_sizes, size_t dimension_count);
-
-  TensorShape(const std::vector<int64_t>& dims, size_t start, size_t end);
+  // Create a TensorShape that points to an existing buffer internally. As no copy is made, 'data' must remain valid for the life of the TensorShape
+  static const TensorShape FromExistingBuffer(const std::vector<int64_t>& data) { return TensorShape(External{}, gsl::span<int64_t>(const_cast<int64_t*>(data.data()), data.size())); }
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/framework/tensor_shape.h:39]: I feel this is a valid concern and with shapes lengths of 5, it is better to copy.

File: include/onnxruntime/core/framework/tensor_shape.h [hunk 2]
```diff
@@ -134,14 +121,22 @@ class TensorShape : private std::vector<int64_t> {
...
+ private:
+
+  struct External {};
+  TensorShape(External, gsl::span<int64_t> buffer) : values_{buffer} {}
+
+  void Allocate(size_t size);
+
+  gsl::span<int64_t> values_;
+  int64_t small_buffer_[5];
+  std::unique_ptr<int64_t[]> allocated_buffer_;
```
Comment (yuslepukhin) [inline include/onnxruntime/core/framework/tensor_shape.h:137]: I feel that we could use std::pmr::vector with pool allocator sitting on top of the buffer with upstream new_delete. That way we perfectly simulate small value optimization with first bytes being taking from the inline buffer and bigger buffers from heap. This way we would get rid of unique_ptr and all the logic associated with it.

Comment (yuslepukhin) [reply include/onnxruntime/core/framework/tensor_shape.h:137]: Furthermore, we could abstract all the logic of manipulating mutable shapes inside TensorShape based on pmr and enjoy correctness and the absence of heap.

Comment (RyanUnderhill) [reply include/onnxruntime/core/framework/tensor_shape.h:137]: We talked about it outside of github, here's a summary:

std::pmr::vector would be useful in all of the code that currently uses TensorShape::GetDimsAsVector to take advantage of a small size optimized vector. As TensorShape isn't a vector (no runtime adding/removing of elements) it's not needed there (and would take up more space).

Sadly there's no way to remove the unique_ptr member as without it there is no way to know if the memory is the small block, allocated memory, or external memory from TensorShape::FromExistingBuffer.

File: include/onnxruntime/core/framework/tensor_shape.h (hunk 1 above)
Comment (pranavsharma) [inline include/onnxruntime/core/framework/tensor_shape.h:39]: nit: can use a new line for the function body

**Review by pranavsharma — COMMENTED** [review]
LGTM 👍

File: include/onnxruntime/core/framework/tensor_shape.h (hunk 1 above)
Comment (RyanUnderhill) [reply include/onnxruntime/core/framework/tensor_shape.h]: Changed, hopefully it'll pass the build tests now.. hitting random breaks unrelated to my change.

**Review by pranavsharma — DISMISSED** [review]

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**